configure-ironic.sh

```diff
@@ -64,20 +27,68 @@ mkdir -p /shared/html
 mkdir -p /shared/ironic_prometheus_exporter
 
 HTPASSWD_FILE=/etc/ironic/htpasswd
+# The user can provide HTTP_BASIC_HTPASSWD and HTTP_BASIC_HTPASSWD_RPC. If
+# - we are running conductor and HTTP_BASIC_HTPASSWD_RPC is not set but HTTP_BASIC_HTPASSWD is,
+#   fallback to use HTTP_BASIC_HTPASSWD for RPC, to preserve compatibility.
+# - we are running combined API and conductor, and HTTP_BASIC_HTPASSWD_RPC is set, use it
+# - we are running combined and HTTP_BASIC_HTPASSWD_RPC is unset, but HTTP_BASIC_HTPASSWD is,
+#   i.e. API is authenticated. We want to authenticate RPC by default, but the user might override.
+#   Then try to infere the authentication strategy and credentials from /auth/ironic-rpc/auth-config.
+#   If not present, then generate a username and password, create the config file and set
+#   HTTP_BASIC_HTPASSWD_RPC
 if [ -n "${HTTP_BASIC_HTPASSWD}" ]; then
     printf "%s\n" "${HTTP_BASIC_HTPASSWD}" >"${HTPASSWD_FILE}"
+
+    if [ -z "${HTTP_BASIC_HTPASSWD_RPC:-}" ]; then
+
+        # When running both API and Conductor in the same container, we'll try to get the credentials
+        # from /auth/ironic-rpc/auth-config if present, or generate it
+        if [ "${IRONIC_DEPLOYMENT}" == "Combined" ]; then
+            # We try to read the credentials from the config file as it is probably mounted read-only,
+            # We cannot modify it. If it is not set to basic, then do not authenticate the RPC. This is
+            # to ensure that the setup will work if the user gives a specific config for rpc set to no_auth
+            if [ -f "/auth/ironic-rpc/auth-config" ]; then
+                IRONIC_RPC_TMP_STRATEGY="$(crudini --get /auth/ironic-rpc/auth-config json-rpc auth_strategy)" || exit 1
+                if [ "${IRONIC_RPC_TMP_STRATEGY}" == "basic" ]; then
+                    IRONIC_RPC_TMP_USERNAME="$(crudini --get /auth/ironic-rpc/auth-config json-rpc username)" || exit 1
+                    IRONIC_RPC_TMP_PASSWORD="$(crudini --get /auth/ironic-rpc/auth-config json-rpc password)" || exit 1
+                fi
+            # We do not have an auth config file, so we generate one
+            else
+                IRONIC_RPC_TMP_USERNAME="$(tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 12 | head -n 1)"
+                IRONIC_RPC_TMP_PASSWORD="$(tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 12 | head -n 1)"
+                mkdir -p "/auth/ironic-rpc"
+                cat << EOF > "/auth/ironic-rpc/auth-config"
```

Comment:
One thing I didn't think about before is that we're writing this to a file that is not on a tmpfs mount (like secrets are), so it could get written to disk at some point. Probably not a reason not to do this, but bindmounting it in from a secret is definitely more secure.

Reply:
I agree that it is not great, but if we want to differentiate the RPC username/password from the API ones, then in the combined case, there is no other option that writing to disk if the user does not provide the credentials. If the user provides the auth-config, then that's more secure.